Source/Core/VideoCommon/PixelEngine.h

```diff
@@ -54,14 +65,157 @@ enum class AlphaReadMode : u16
+  InvSrcAlpha = 5,
+  DstAlpha = 6,
+  InvDstAlpha = 7
+};
 
-void RegisterMMIO(MMIO::Mapping* mmio, u32 base);
+enum class DstBlendFactor : u16
+{
+  Zero = 0,
+  One = 1,
+  SrcClr = 2,
+  InvSrcClr = 3,
+  SrcAlpha = 4,
+  InvSrcAlpha = 5,
+  DstAlpha = 6,
+  InvDstAlpha = 7
+};
 
-// gfx backend support
-void SetToken(const u16 token, const bool interrupt, int cycle_delay);
-void SetFinish(int cycle_delay);
-AlphaReadMode GetAlphaReadMode();
+enum class LogicOp : u16
+{
+  Clear = 0,
+  And = 1,
+  AndReverse = 2,
+  Copy = 3,
+  AndInverted = 4,
+  NoOp = 5,
+  Xor = 6,
+  Or = 7,
+  Nor = 8,
+  Equiv = 9,
+  Invert = 10,
+  OrReverse = 11,
+  CopyInverted = 12,
+  OrInverted = 13,
+  Nand = 14,
+  Set = 15
+};
+
+union UPEAlphaConfReg
+{
+  u16 hex;
+  BitField<0, 1, bool, u16> blend;  // Set for GX_BM_BLEND or GX_BM_SUBTRACT
+  BitField<1, 1, bool, u16> logic;  // Set for GX_BM_LOGIC
+  BitField<2, 1, bool, u16> dither;
+  BitField<3, 1, bool, u16> color_update_enable;
+  BitField<4, 1, bool, u16> alpha_update_enable;
+  BitField<5, 3, DstBlendFactor, u16> dst_factor;
+  BitField<8, 3, SrcBlendFactor, u16> src_factor;
+  BitField<11, 1, bool, u16> subtract;  // Set for GX_BM_SUBTRACT
+  BitField<12, 4, LogicOp, u16> logic_op;
+};
+
+union UPEDstAlphaConfReg
+{
+  u16 hex;
+  BitField<0, 8, u8, u16> alpha;
+  BitField<8, 1, bool, u16> enable;
+};
+
+union UPEAlphaModeConfReg
+{
+  u16 hex;
+  BitField<0, 8, u8, u16> threshold;
+  // Yagcd and libogc use 8 bits for this, but the enum only needs 3
+  BitField<8, 3, CompareMode, u16> compare_mode;
+};
+
+union UPEAlphaReadReg
+{
+  u16 hex;
+  BitField<0, 2, AlphaReadMode, u16> read_mode;
+};
+
+// fifo Control Register
+union UPECtrlReg
+{
+  u16 hex;
+  BitField<0, 1, bool, u16> pe_token_enable;
+  BitField<1, 1, bool, u16> pe_finish_enable;
+  BitField<2, 1, bool, u16> pe_token;   // Write only
+  BitField<3, 1, bool, u16> pe_finish;  // Write only
+};
+
+class PixelEngineManager
+{
+public:
+  void Init(Core::System& system);
+  void DoState(PointerWrap& p);
+
+  void RegisterMMIO(MMIO::Mapping* mmio, u32 base);
+
+  // gfx backend support
+  void SetToken(Core::System& system, const u16 token, const bool interrupt, int cycle_delay);
+  void SetFinish(Core::System& system, int cycle_delay);
+  AlphaReadMode GetAlphaReadMode() const { return m_alpha_read.read_mode; }
+
+private:
+  void RaiseEvent(Core::System& system, int cycles_into_future);
+  void UpdateInterrupts();
+  void SetTokenFinish_OnMainThread(Core::System& system, u64 userdata, s64 cycles_late);
+
+  static void SetTokenFinish_OnMainThread_Static(Core::System& system, u64 userdata,
+                                                 s64 cycles_late);
+
+  // STATE_TO_SAVE
+  UPEZConfReg m_z_conf;
+  UPEAlphaConfReg m_alpha_conf;
+  UPEDstAlphaConfReg m_dst_alpha_conf;
+  UPEAlphaModeConfReg m_alpha_mode_conf;
+  UPEAlphaReadReg m_alpha_read;
+  UPECtrlReg m_control;
+
+  std::mutex m_token_finish_mutex;
+  u16 m_token;
```

Comment:
This should all have default initializers, so they're all consistently initialized on construction, even before `Init` is called

Reply:
Oops, yes, fixed.